### DISCUSSION TITLE
Which kernel can model U_xy term

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#19475] [Q&A General] [answered]
Hi

Given the following differential equation, which kernel should I use to model the second and the third terms?
![image](https://user-images.githubusercontent.com/88537645/143295812-1aa1c5e0-2096-4b5d-b42f-71277d19fafe.png)

Or, I would need to implement a custom one?

**Comment by GiudGiud** [comment]
Hello

I dont know of one but libmesh can compute hessians so it should be possible. @lindsayad do we have this in MOOSE? I had no luck looking in the variable code. 

Guillaume

**Comment by lindsayad** [comment]
It's possible that somewhere in the MOOSE app eco-system someone has this, but I don't believe we have this in the MOOSE repository itself. So you will likely need to implement a custom one. Is `u` your nonlinear variable for the equation you posted? If so you can couple in second derivatives of variables. You can look [here](https://github.com/idaholab/moose/blob/next/modules/navier_stokes/src/kernels/INSBase.C#L63) for an example of using the `coupledSecond` API. The automatic differentiation (AD) version is `adCoupledSecond`. I believe the tensor that is returned to you will be upper triangular with the knowledge that partial derivatives commute. So you should always index u_ij with j >= i, I believe. If you don't, you'll probably always get zeroes returned. @roystgnr could potentially weigh in on this last bit

**Reply by lindsayad** [reply]
@jwpeterson would also know about the libMesh hessians and if my reading of the code is right and it's upper triangular, then our current SUPG stabilization for INS is not quite right, e.g. this code should not be as simple as it is now
```c++
RealVectorValue
INSBase::strongViscousTermTraction()
{
  return strongViscousTermLaplace() -
         _mu[_qp] *
             (_second_u_vel[_qp].row(0) + _second_v_vel[_qp].row(1) + _second_w_vel[_qp].row(2));
}
```

**Reply by roystgnr** [reply]
You should always see `u_ij == u_ji`; internally we loop over only one triangle's entries when computing basis second derivatives, but we copy the off-diagonal entries into the other triangle too.

In hindsight I should have created a SymmetricTensor class, though.  You can write simple kernels thanks to the full tensor being complete, but generally as soon as that's done you want to rewrite them to also only loop over one triangle's entries to again save a few flops.

**Reply by jwpeterson** [reply]
@lindsayad The code you have above looks correct to me. In Cartesian coords and using index notation, the term in question is:
```
u_i,jj + u_j,ij
```
with summation over the dummy index j, and where the comma represents differentiation. The `strongViscousTermLaplace()` call gives you the first term, and I checked that the various `row` calls you make give you the the second term. In cylindrical coords the second derivatives are a bit messier... not sure if we claim to support that.

**Reply by lindsayad** [reply]
Oh what a dope. I only looking at the comment in the h1 transformation code. Right there in the equalities ... both xy and yx
```c++
              // phi_{x y}
              d2phi[i][p].slice(0).slice(1) = d2phi[i][p].slice(1).slice(0) = d2phidxdy[i][p] =
```

**Reply by aaelmeli** [reply]
> d I checked that the various `ro`

Yesm

> It's possible that somewhere in the MOOSE app eco-system someone has this, but I don't believe we have this in the MOOSE repository itself. So you will likely need to implement a custom one. Is `u` your nonlinear variable for the equation you posted? If so you can couple in second derivatives of variables. You can look [here](https://github.com/idaholab/moose/blob/next/modules/navier_stokes/src/kernels/INSBase.C#L63) for an example of using the `coupledSecond` API. The automatic differentiation (AD) version is `adCoupledSecond`. I believe the tensor that is returned to you will be upper triangular with the knowledge that partial derivatives commute. So you should always index u_ij with j >= i, I believe. If you don't, you'll probably always get zeroes returned. @roystgnr could potentially weigh in on this last bit

Yes, U and W are my nonlinear variables that are coupled and live on the same mesh.

**Comment by jwpeterson** [comment, marked as answer]
@aaelmeli This is a strong form PDE, once you multiply by a test function and apply the divergence theorem, you won't need to use second derivatives, right? Unless I guess you are planning to do some sort of stabilization scheme that requires the strong form...

**Reply by aaelmeli** [reply]
Yes, this is the strong form PDE. As you said, I won't need to use the second derivative after forming the weak form, however, I am not aware of a kernel in moose that can model the second term.

**Reply by lindsayad** [reply]
> aware of a kernel in moose that can model the second term.

Yea I don't think we have anything for you. How does this equation generalize to 3D? I'm curious whether there is a nice way to write this with divergence and gradient operators.

**Reply by aaelmeli** [reply]
In 3D, after some manipulation to the coefficient, this can be written as 
![image](https://user-images.githubusercontent.com/88537645/144115300-29090c9c-eb4c-4be2-a933-2840560dc60f.png)

Yes, that is what I am trying. I am trying to use stress divergence to model this PDE.

**Reply by jwpeterson** [reply]
Your 3D PDE is in the usual divergence form
```
div \sigma(u) = b
```
where \sigma is a stress tensor, so the weak form is also standard,
```
-\int \sigma(u) : grad test = \int b . test + boundary terms
```
Not sure if you need to introduce the auxiliary variables "theta" and "e", but it should be possible to do in MOOSE.
